File: internal/controllers/qr_code_controller.go
package controllers

import (
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/growthOS/qr-review-service/internal/dto"
	"github.com/growthOS/qr-review-service/internal/repositories"
	"github.com/growthOS/qr-review-service/internal/services"
	"github.com/growthOS/qr-review-service/pkg/qrgen"
)

// QRCodeController handles HTTP requests for QR code operations.
type QRCodeController struct {
	service  *services.QRCodeService
	qrRepo   repositories.QRCodeRepository
	shopRepo repositories.ShopRepository
	baseURL  string
}

// NewQRCodeController creates a new QR code controller.
func NewQRCodeController(service *services.QRCodeService, qrRepo repositories.QRCodeRepository, shopRepo repositories.ShopRepository, baseURL string) *QRCodeController {
	return &QRCodeController{service: service, qrRepo: qrRepo, shopRepo: shopRepo, baseURL: baseURL}
}

// Create handles POST /api/v1/qr-reviews/qr
func (ctrl *QRCodeController) Create(c *gin.Context) {
	var req dto.CreateQRCodeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"message": "Invalid request: " + err.Error(),
		})
		return
	}

	qrResp, err := ctrl.service.CreateQRCode(req)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"message": err.Error(),
		})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"success": true,
		"data":    qrResp,
		"message": "QR code created successfully",
	})
}

// BulkCreate handles POST /api/v1/qr-reviews/qr/bulk — pre-generates unlinked QR codes.
func (ctrl *QRCodeController) BulkCreate(c *gin.Context) {
	var req dto.BulkCreateQRCodeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"message": "Invalid request: " + err.Error(),
		})
		return
	}

	results, err := ctrl.service.BulkCreateQRCodes(req)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"message": err.Error(),
		})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"success": true,
		"data":    results,
		"message": fmt.Sprintf("%d QR codes created successfully", len(results)),
	})
}

// Activate handles POST /api/v1/qr-reviews/qr/:id/activate — links a QR code to a business.
func (ctrl *QRCodeController) Activate(c *gin.Context) {
	qrID := c.Param("id")

	var req dto.ActivateQRCodeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"message": "Invalid request: " + err.Error(),
		})
		return
	}

	shop, err := ctrl.service.ActivateQRCode(qrID, req)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"message": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "QR code activated successfully",
		"data": gin.H{
			"qr_id":   qrID,
			"shop_id":  shop.ID.String(),
			"shop_name": shop.Name,
		},
	})
}

// GetByID handles GET /api/v1/qr-reviews/qr/:id
func (ctrl *QRCodeController) GetByID(c *gin.Context) {
	id := c.Param("id")

	qrResp, err := ctrl.service.GetQRCodeByID(id)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"success": false,
			"message": "QR code not found",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"data":    qrResp,
	})
}

// Image handles GET /qr-image/:id — serves the QR code as a PNG image.
func (ctrl *QRCodeController) Image(c *gin.Context) {
	id := c.Param("id")
	qrCode, err := ctrl.qrRepo.FindByID(id)
	if err != nil || qrCode == nil {
		c.String(http.StatusNotFound, "QR code not found")
		return
	}

	qrURL := fmt.Sprintf("%s/r/%s", ctrl.baseURL, qrCode.ID)
	png, err := qrgen.GenerateBytes(qrURL, 512)
	if err != nil {
		c.String(http.StatusInternalServerError, "Failed to generate QR image")
		return
	}

	c.Data(http.StatusOK, "image/png", png)
}

// Dashboard handles GET /dashboard/:shop_id — shows all QR codes for a shop.
func (ctrl *QRCodeController) Dashboard(c *gin.Context) {
	shopIDStr := c.Param("shop_id")
	shopID, err := uuid.Parse(shopIDStr)
	if err != nil {
		c.HTML(http.StatusBadRequest, "dashboard.html", gin.H{"Error": "Invalid shop ID"})
		return
	}

	shop, err := ctrl.shopRepo.FindByID(shopID)
	if err != nil {
		c.HTML(http.StatusNotFound, "dashboard.html", gin.H{"Error": "Shop not found"})
		return
	}

	qrCodes, err := ctrl.qrRepo.FindByShopID(shopID)
	if err != nil {
		c.HTML(http.StatusInternalServerError, "dashboard.html", gin.H{"Error": "Failed to load QR codes"})
		return
	}

	type qrData struct {
		ID        string
		Label     string
		ScanCount int
		ImageURL  string
		ScanURL   string
	}
	var codes []qrData
	for _, qr := range qrCodes {
		codes = append(codes, qrData{
			ID:        qr.ID,
			Label:     qr.Label,
			ScanCount: qr.ScanCount,
			ImageURL:  fmt.Sprintf("/qr-image/%s", qr.ID),
			ScanURL:   fmt.Sprintf("%s/r/%s", ctrl.baseURL, qr.ID),
		})
	}

	c.HTML(http.StatusOK, "dashboard.html", gin.H{
		"ShopName": shop.Name,
		"ShopID":   shop.ID.String(),
		"QRCodes":  codes,
	})
}

File: backend/src/controllers/qrCodeController.js
const qrCodeService = require("../services/qrCodeService");
const prisma = require("../lib/prisma");
const { generateBytes } = require("../lib/qrgen");
const config = require("../config/config");
const { parsePagination, paginationMeta } = require("../lib/pagination");

async function create(req, res) {
  try {
    const qrResp = await qrCodeService.createQRCode(req.body || {});
    res.status(201).json({
      success: true,
      data: qrResp,
      message: "QR code created successfully",
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
}

async function bulkCreate(req, res) {
  const { count } = req.body || {};
  if (!count || count < 1 || count > 500) {
    return res.status(400).json({
      success: false,
      message: "Invalid request: count must be between 1 and 500",
    });
  }

  try {
    const results = await qrCodeService.bulkCreateQRCodes(req.body);
    res.status(201).json({
      success: true,
      data: results,
      message: `${results.length} QR codes created successfully`,
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
}

async function activate(req, res) {
  const qrId = req.params.id;
  const { business_name, review_url } = req.body || {};
  if (!business_name || !review_url) {
    return res.status(400).json({
      success: false,
      message: "Invalid request: business_name and review_url are required",
    });
  }

  try {
    const shop = await qrCodeService.activateQRCode(qrId, req.body);
    res.status(200).json({
      success: true,
      message: "QR code activated successfully",
      data: {
        qr_id: qrId,
        shop_id: shop.id,
        shop_name: shop.name,
      },
    });
  } catch (err) {
    res.status(400).json({ success: false, message: err.message });
  }
}

async function getById(req, res) {
  try {
    const qrResp = await qrCodeService.getQRCodeById(req.params.id);
    res.status(200).json({ success: true, data: qrResp });
  } catch (err) {
    res.status(404).json({ success: false, message: "QR code not found" });
  }
}

async function image(req, res) {
  const id = req.params.id;
  try {
    const qrCode = await prisma.qRCode.findUnique({ where: { id } });
    if (!qrCode) {
      return res.status(404).send("QR code not found");
    }

    const qrUrl = `${config.frontendUrl}/r/${qrCode.id}`;
    const png = await generateBytes(qrUrl, 512);
    res.set("Content-Type", "image/png");
    res.status(200).send(png);
  } catch (err) {
    res.status(500).send("Failed to generate QR image");
  }
}

// dashboard returns JSON data for all QR codes belonging to a shop (consumed by the React frontend).
async function dashboard(req, res) {
  const shopId = req.params.shop_id;

  try {
    const shop = await prisma.shop.findUnique({ where: { id: shopId } });
    if (!shop) {
      return res.status(404).json({ success: false, message: "Shop not found" });
    }

    if (req.user.role === "OWNER" && shop.ownerUserId !== req.user.id) {
      return res.status(403).json({ success: false, message: "Forbidden: not your shop" });
    }

    const { page, limit, skip } = parsePagination(req.query);

    const [qrCodes, total] = await Promise.all([
      prisma.qRCode.findMany({
        where: { shopId },
        orderBy: { createdAt: "desc" },
        skip,
        take: limit,
      }),
      prisma.qRCode.count({ where: { shopId } }),
    ]);

    const codes = qrCodes.map((qr) => ({
      id: qr.id,
      label: qr.label,
      scan_count: qr.scanCount,
      image_url: `${config.baseUrl}/qr-image/${qr.id}`,
      scan_url: `${config.frontendUrl}/r/${qr.id}`,
    }));

    res.status(200).json({
      success: true,
      data: {
        shop_name: shop.name,
        shop_id: shop.id,
        qr_codes: codes,
      },
      meta: paginationMeta(page, limit, total),
    });
  } catch (err) {
    res.status(500).json({ success: false, message: "Failed to load QR codes" });
  }
}

// listAll returns every QR code with its linked shop (if any) — admin/salesman only.
async function listAll(req, res) {
  try {
    const { page, limit, skip } = parsePagination(req.query);

    const [qrCodes, total] = await Promise.all([
      prisma.qRCode.findMany({
        orderBy: { createdAt: "desc" },
        include: { shop: { select: { id: true, name: true } } },
        skip,
        take: limit,
      }),
      prisma.qRCode.count(),
    ]);

    const results = qrCodes.map((qr) => ({
      id: qr.id,
      label: qr.label,
      scan_count: qr.scanCount,
      is_active: qr.isActive,
      is_linked: !!qr.shopId,
      shop_id: qr.shopId || "",
      shop_name: qr.shop?.name || "",
      scan_url: `${config.frontendUrl}/r/${qr.id}`,
      created_at: qr.createdAt.toISOString(),
    }));

    res.status(200).json({ success: true, data: results, meta: paginationMeta(page, limit, total) });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
}

module.exports = { create, bulkCreate, activate, getById, image, dashboard, listAll };
